freqtrade/data/history/history_utils.py

```diff
@@ -152,71 +152,90 @@ def _load_cached_data_for_updating(pair: str, timeframe: str, timerange: Optiona
+
+        cached_start, cached_end = None, None
+        if not cached.empty:
+            cached_start = int(cached.iloc[0]['date'].timestamp())
+            cached_end = int(cached.iloc[-1]['date'].timestamp())
+
+        logger.debug("Cached Start: %s",
+                     f"{cached.iloc[0]['date']:%Y-%m-%d %H:%M:%S}"
+                     if cached_start else 'None')
+        logger.debug("Cached End: %s",
+                     f"{cached.iloc[-1]['date']:%Y-%m-%d %H:%M:%S}"
+                     if cached_end else 'None')
+
+        # Set the bounds for downloading
+        since_ms, until_ms = None, None
+        if cached.empty:
+            since_ms = (since if since else
+                        arrow.utcnow().shift(days=-30).int_timestamp) * 1000
+            until_ms = (until if until else
+                        arrow.utcnow().int_timestamp) * 1000
+        else:
+            # Determine lower bound
+            if since:
+                since_ms = (since if since < cached_end else cached_end) * 1000
+            else:
+                # Set lower-bound to the end of the cache if not provided
+                since_ms = cached_end * 1000
+
+            # Determine upper bound
+            if until:
+                until_ms = (until if until > cached_start else cached_start) \
+                           * 1000
+            else:
+                # Default upper-bound is the current time.
+                until_ms = arrow.utcnow().int_timestamp * 1000
 
-        # Default since_ms to 30 days if nothing is given
         new_data = exchange.get_historic_ohlcv(pair=pair,
                                                timeframe=timeframe,
-                                               since_ms=since_ms if since_ms else
-                                               int(arrow.utcnow().shift(
-                                                   days=-30).float_timestamp) * 1000
-                                               )
-        # TODO: Maybe move parsing to exchange class (?)
-        new_dataframe = ohlcv_to_dataframe(new_data, timeframe, pair,
-                                           fill_missing=False, drop_incomplete=True)
-        if data.empty:
-            data = new_dataframe
+                                               since_ms=since_ms,
+                                               until_ms=until_ms)
+
+        new_df = ohlcv_to_dataframe(new_data, timeframe, pair,
+                                    fill_missing=False)
+
+        if cached.empty:
+            data = new_df
         else:
-            # Run cleaning again to ensure there were no duplicate candles
-            # Especially between existing and new data.
-            data = clean_ohlcv_dataframe(data.append(new_dataframe), timeframe, pair,
-                                         fill_missing=False, drop_incomplete=False)
+            data = clean_ohlcv_dataframe(cached.append(new_df), timeframe,
```

Comment:
I would keep the comment we had before this line - as it makes it clear what `clean_ohlcv_dataframe()` does (i had to look at that function to see that it'll remove duplicates)...